src/java/org/apache/cassandra/index/sai/utils/IndexTermType.java

```diff
@@ -0,0 +1,858 @@
+import org.apache.cassandra.db.marshal.StringType;
+import org.apache.cassandra.db.marshal.UTF8Type;
+import org.apache.cassandra.db.marshal.UUIDType;
+import org.apache.cassandra.db.marshal.VectorType;
+import org.apache.cassandra.db.rows.Cell;
+import org.apache.cassandra.db.rows.ComplexColumnData;
+import org.apache.cassandra.db.rows.Row;
+import org.apache.cassandra.index.sai.plan.Expression;
+import org.apache.cassandra.schema.ColumnMetadata;
+import org.apache.cassandra.serializers.MarshalException;
+import org.apache.cassandra.utils.ByteBufferUtil;
+import org.apache.cassandra.utils.FastByteOperations;
+import org.apache.cassandra.utils.bytecomparable.ByteComparable;
+import org.apache.cassandra.utils.bytecomparable.ByteSource;
+import org.apache.cassandra.utils.bytecomparable.ByteSourceInverse;
+
+/**
+ * This class is a representation of an {@link AbstractType} as an indexable type. It is responsible for determining the
+ * capabilities of the type and provides helper methods for handling term values associated with the type.
+ */
+public class IndexTermType
+{
+    private static final Set<AbstractType<?>> EQ_ONLY_TYPES = ImmutableSet.of(UTF8Type.instance,
+                                                                              AsciiType.instance,
+                                                                              BooleanType.instance,
+                                                                              UUIDType.instance);
+
+    private static final byte[] IPV4_PREFIX = new byte[] { 0, 0, 0, 0, 0, 0, 0, 0, 0, 0, -1, -1 };
+
+    /**
+     * DecimalType / BigDecimal values are indexed by truncating their asComparableBytes representation to this size,
+     * padding on the right with zero-value-bytes until this size is reached (if necessary).  This causes
+     * false-positives that must be filtered in a separate step after hitting the index and reading the associated
+     * (full) values.
+     */
+    private static final int DECIMAL_APPROXIMATION_BYTES = 24;
+    private static final int BIG_INTEGER_APPROXIMATION_BYTES = 20;
+    private static final int INET_ADDRESS_SIZE = 16;
+    private static final int DEFAULT_FIXED_LENGTH = 16;
+
+    private static final int IS_LITERAL               = 1;
+    private static final int IS_STRING                = 1 << 1;
+    private static final int IS_VECTOR                = 1 << 2;
+    private static final int IS_REVERSED              = 1 << 3;
+    private static final int IS_FROZEN                = 1 << 4;
+    private static final int IS_NON_FROZEN_COLLECTION = 1 << 5;
+    private static final int IS_COMPOSITE             = 1 << 6;
+    private static final int IS_INET_ADDRESS          = 1 << 7;
+    private static final int IS_BIG_INTEGER           = 1 << 8;
+    private static final int IS_BIG_DECIMAL           = 1 << 9;
+    private static final int IS_LONG                  = 1 << 10;
+    private static final int IS_COMPOSITE_PARTITION   = 1 << 11;
```

Comment:
Any reason not to use `EnumSet` here? Something like...

```
enum Flag { IS_LONG, IS_COMPOSITE_PARTITION, ... }
private final EnumSet<Flag> capabilities = EnumSet.of(Flag.IS_LONG, Flag.IS_COMPOSITE_PARTITION);
```

Reply:
...and usage...

```
private boolean isCompositePartition()
    {
        return capabilities.contains(Flag.IS_COMPOSITE_PARTITION);
    }
```

Reply:
I had thought about doing this but (for some reason) stuck with the bitset. I think an `EnumSet` would be far better for debugging, I have been finding the bitset a pain. I'll get that in now.